cartridge/upload.lua

```diff
@@ -0,0 +1,322 @@
+--- Spread the data across instances in a network-efficient manner.
+--
+-- (**Added** in v2.4.0-43)
+--
+-- @module cartridge.upload
+-- @local
+
+local log = require('log')
+local fio = require('fio')
+local errno = require('errno')
+local fiber = require('fiber')
+local checks = require('checks')
+local errors = require('errors')
+local digest = require('digest')
+local msgpack = require('msgpack')
+local pool = require('cartridge.pool')
+local utils = require('cartridge.utils')
+
+local vars = require('cartridge.vars').new('cartridge.upload')
+
+local UploadError = errors.new_class('UploadError')
+
+--- The uploaded data.
+-- `{[upload_id] = data}`
+-- @table inbox
+vars:new('inbox', {})
+
+-- All four functions begin / transmit / finish / cleanup do yield.
+-- Since they are called over netbox (each in an individual fiber),
+-- there may be a situation (usually abnormal), when cleanup starts
+-- before previous stages finish.
+vars:new('upload_fibers', {})
+
+-- Temporary directory used for saving files during upload.
+vars:new('upload_prefix', '/tmp')
+
+local function set_upload_prefix(prefix)
+    vars.upload_prefix = prefix
+end
+
+local function get_upload_prefix()
+    return vars.upload_prefix
+end
+
+local function get_upload_path(upload_id)
+    checks('string')
+    return fio.pathjoin(
+        vars.upload_prefix,
+        'cartridge-upload.' .. upload_id
+    )
+end
+
+-- The upload starts by creating a shared resource - a directory named
+-- after the `upload_id`. The instances who can create it before
+-- others are the transmitters. The others wait patiently till the
+-- `upload_finish()`.
+local function upload_begin(upload_id)
+    checks('string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local ok, err = utils.mktree(fio.dirname(upload_path))
+    fiber.testcancel()
+    if ok == nil then
+        vars.upload_fibers[upload_id] = nil
+        return nil, err
+    end
+
+    local ok = fio.mkdir(upload_path)
+    local _errno = errno()
+    fiber.testcancel()
+
+    vars.upload_fibers[upload_id] = nil
+
+    if ok then
+        return true
+    elseif fio.path.is_dir(upload_path) then
+        return false
+    else
+        return nil, UploadError:new(
+            'Error creating directory %q: %s',
+            upload_path, errno.strerror(_errno)
+        )
+    end
+end
+
+-- The communication only goes with transmitters - the instances who
+-- replied `upload_begin() == true` i.e affirmed they own the shared
+-- resource. The other instances will read the data from that resource
+-- during the `upload_finish()` step.
+local function upload_transmit(upload_id, payload)
+    checks('string', 'string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local payload_path = fio.pathjoin(upload_path, 'payload')
+    local ok, err = utils.file_write(payload_path, payload)
+    fiber.testcancel()
```

Comment:
`testcancel` raises. You won't perform `vars.upload_fibers[upload_id] = nil` if this function raises an error.